src/main/java/net/spy/memcached/tls/TLSConnectionManager.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    private void configureSslEngine() {
+        // We are the client, not the server
+        sslEngine.setUseClientMode(true);
+    }
+
+    private void initBuffers(SSLSession session) {
+        // allocateDirect() to keep all bytes contiguous in memory
+        appOutBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        appInBuffer = ByteBuffer.allocateDirect(session.getApplicationBufferSize());
+        networkOutBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+        networkInBuffer = ByteBuffer.allocateDirect(session.getPacketBufferSize());
+    }
+
+    public boolean doHandshake(SocketChannel socketChannel) throws IOException {
+        LOG.debug("%s - Beginning handshake.", socketChannel.getRemoteAddress());
+        try {
+            sslEngine.beginHandshake();
+            currentSession = sslEngine.getSession();
+            initBuffers(currentSession);
+
+            HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+            while (!Thread.currentThread().isInterrupted()
+                    && handshakeStatus != HandshakeStatus.FINISHED
+                    && handshakeStatus != HandshakeStatus.NOT_HANDSHAKING) {
+                LOG.debug("%s - Handshake status: %s", socketChannel.getRemoteAddress(), handshakeStatus.name());
+                switch (handshakeStatus) {
+                    case NEED_TASK:
+                        // we need to finish these tasks for the handshake to continue
+                        List<Future<?>> tasks = new ArrayList<>();
+                        Runnable currentTask = sslEngine.getDelegatedTask();
+                        while(currentTask != null) {
+                            tasks.add(executor.submit(currentTask));
+                            currentTask = sslEngine.getDelegatedTask();
+                        }
+                        for (Future<?> task : tasks) {
+                            try {
+                                task.get();
+                            } catch (ExecutionException e) {
+                                // TODO make this more clear
+                                throw new RuntimeException(e);
+                            }
+                        }
+                        break;
+                    case NEED_WRAP:
+                        // We're about to send something to the server, but it needs to be wrapped first
+                        networkOutBuffer.clear();
+                        SSLEngineResult wrapResult = sslEngine.wrap(appOutBuffer, networkOutBuffer);
+                        handleHandshakeWrapResult(socketChannel, wrapResult);
+                        break;
+                    case NEED_UNWRAP:
+                        // We need to receive a message from the server, but it needs to be unwrapped first
+                        if (socketChannel.read(networkInBuffer) < 0) {
+                            // The message was empty
```

Comment:
this one seems like a good candidate for a helper method so you can do like `if(messageFromServerIsEmpty)` instead of a comment